src/src/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -0,0 +1,359 @@
+        if (containOauthError(mrrtResult)) {
+            // If MRRT fails, we still want to retry on FRT in case there is one there. 
+            // MRRT may not be marked as FRT, hard-code it as "1" in this case. 
+            final String familyClientId = StringExtensions.IsNullOrBlank(mMrrtTokenCacheItem.getFamilyClientId()) ? 
+                    AuthenticationConstants.MS_FAMILY_ID : mMrrtTokenCacheItem.getFamilyClientId();
+
+            // Rest mMrrtTokenCacheItem to avoid the fallback when FRT request fails.
+            mMrrtTokenCacheItem = null;
+            mrrtResult = tryFRT(familyClientId, mrrtResult);
+        }
+        
+        return mrrtResult;
+    }
+    
+    /**
+     * Attempt to get access token with FRT. 
+     */
+    private AuthenticationResult tryFRT(final String familyClientId, final AuthenticationResult mrrtResult) 
+            throws AuthenticationException {
+        if (mAttemptedWithFRT) {
+            return mrrtResult;
+        }
+        
+        mAttemptedWithFRT = true;
+        final TokenCacheItem frtTokenCacheItem = mTokenCacheAccessor.getFRTItem(familyClientId, getUser());
+        
+        if (frtTokenCacheItem  == null) {
+            if (mMrrtTokenCacheItem != null) {
+                // if mrrt token cache item is already used for token request, it will be reset as null
+                Logger.v(TAG, "FRT cache item does not exist, fall back to try MRRT.");
+                return tryMRRT();
+            } else {
+                // No FRT existed, and we've already tried to get new AT with MRRT, returning MRRT token request result in this case. 
+                return mrrtResult;
+            }
+        }
+        
+        Logger.v(TAG, "Send request to use FRT for new AT.");
+        AuthenticationResult frtResult = acquireTokenWithCachedItem(frtTokenCacheItem);
+        if (containOauthError(frtResult)) {
+            final AuthenticationResult retryMrrtResult = tryMRRT();
+            frtResult = retryMrrtResult == null ? frtResult : retryMrrtResult;
+        }
+        
+        return frtResult;
+    }
+    
+    /**
+     * Acquire token with retrieved token cache item and update cache. 
+     */
+    private AuthenticationResult acquireTokenWithCachedItem(final TokenCacheItem cachedItem)
+            throws AuthenticationException {
+        final AuthenticationResult result = acquireTokenWithRefreshToken(cachedItem.getRefreshToken());
+        mTokenCacheAccessor.updateCachedItemToResult(mAuthRequest.getResource(), mAuthRequest.getClientId(), 
+                result, cachedItem);
+
+        return result;
+    }
+    
+    /**
+     * Old version of ADAL doesn't mark token stored in regular RT entry as MRRT even it is. The logic to look for 
+     * MRRT is when RT is not found or found RT is also MRRT. To support the old behavior, do a separate check on
+     * the existence for MRRT token entry. 
+     */
+    private boolean isMRRTExisted() {
```

Comment:
Done. 

---
In reply to: [62922142](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/611#discussion_r62922142) [](ancestors = 62922142)